pa1/PPToken.h

```diff
@@ -0,0 +1,190 @@
+#ifndef PPToken_h
+#define PPToken_h
+
+#include <memory>
+#include <string>
+#include <vector>
+
+// This file defines the interface of PPToken and its subclasses.
+
+////////////////////////////////////////////////////////////////////////////////
+// Token type class
+////////////////////////////////////////////////////////////////////////////////
+enum class PPTokenType: int {
+  // 9 token types defined in the C++ spec
+  HeaderName = 0,
+  Identifier,
+  PPNumber,
+  CharacterLiteral,
+  UserDefinedCharacterLiteral,
+  StringLiteral,
+  UserDefinedStringLiteral,
+  PreprocessingOpOrPunc,
+  NonWhitespaceChar,
+
+  // Additional token types useful for preprocessing
+  //
+  // new_line: self-exaplaining
+  //
+  // whitespace_sequence:
+  //    maximal sequence of whitespace characters not including line feeds,
+  //    (which effectively also includes comments)
+  NewLine,
+  WhitespaceSequence,
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// Forward declarations of subclasses
+////////////////////////////////////////////////////////////////////////////////
+class PPTokenHeaderName;
+class PPTokenIdentifier;
+class PPTokenPPNumber;
+class PPTokenCharacterLiteral;
+class PPTokenUserDefinedCharacterLiteral;
+class PPTokenStringLiteral;
+class PPTokenUserDefinedStringLiteral;
+class PPTokenPreprocessingOpOrPunc;
+class PPTokenNonWhitespaceChar;
+class PPTokenNewLine;
+class PPTokenWhitespaceSequence;
+
+////////////////////////////////////////////////////////////////////////////////
+// Abstract base class
+////////////////////////////////////////////////////////////////////////////////
+class PPToken {
+public:
+  PPToken(const PPTokenType type): _type(type) {}
+
+  // Interface
+  PPTokenType getType() const { return _type; }
+
+  // Get the corresponding raw text
+  virtual std::string getUTF8String() const = 0;
+
+  // Get human friendly UTF8 string for the given token type
```

Comment:
Should probably document why this function is `static`:

While at first glance it is better to write a non-static pure virtual method `std::string getTypeUTF8String() const = 0` to `PPCodeUnit`, it is indeed much clearer to just list all the preprocessing token types together in a list in the .cpp file. Given that I am using two custom token types, `new-line` and `whitespace-sequence`, this was the least error-prone way of having this functionality when I first developed this structure because I can just copy and paste the list of token types from the C++ specification PDF and know it is guaranteed to be right. Not to mention it saves a lot of coding, compared to having separate methods in each subclass.